### PR TITLE
Add Kokkos version of spica pair and angle styles

### PR DESCRIPTION
**Pull request by alphataubio**
**Summary**

Add pair_style lj/spica/coul/long/kk and angle_style spica/kk

**Related Issue(s)**

n/a

**Author(s)**

Mitch Murphy
alphataubio@gmail.com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

n/a

**Implementation Notes**

- pair_lj_spica_coul_long_kokkos based on pair_lj_spica_kokkos and pair_lj_charmmfsw_coul_long_kokkos
- angle_spica_kokkos based on angle_harmonic_kokkos

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by alphataubio** [comment]
i added pair and angle styles to KOKKOS/Install.sh but i wasnt able to get conventional make to work with openmp on macos or linux

**Comment by alphataubio** [comment]
i merged branch 'lammps:develop' into spica-kk and now unit test is failing in 
projectroot.unittest.force-styles.AtomicPairStyle:table_bitmap

```
Stacktrace
--------------------------------------------------------------------------
A call to mkdir was unable to create the desired directory:

  Directory: /tmp/ompi.atlas2.1003
  Error:     File exists

Please check to ensure you have adequate permissions to perform
the desired operation.
--------------------------------------------------------------------------
[atlas2:01082] [[0,1],0] ORTE_ERROR_LOG: Error in file util/session_dir.c at line 107
[atlas2:01082] [[0,1],0] ORTE_ERROR_LOG: Error in file util/session_dir.c at line 346
[atlas2:01082] [[0,1],0] ORTE_ERROR_LOG: Error in file ess_singleton_module.c at line 340
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  orte_session_dir failed
  --> Returned value Error (-1) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
```

**Comment by akohlmey** [comment]
> i merged branch 'lammps:develop' into spica-kk and now unit test is failing in
> projectroot.unittest.force-styles.AtomicPairStyle:table_bitmap

This has nothing to do with you. This is a side effect of a race condition in the OpenMPI library version used in the Ubuntu 20.04LTS containers that are used in the automated testing. We have made some changes to avoid this as much as possible, but every once in a while it still gets triggered. In all likelihood this will go away at the next push to the branch.

**Comment by stanmoore1** [comment]
@alphataubio @alphataubio I've reviewed the code in this PR and have a few minor change requests/comments, will submit soon. But otherwise it looks good I think it can be merged soon. @alphataubio Also is there an example I can use to test the new styles?

File: src/KOKKOS/pair_lj_spica_coul_long_kokkos.cpp [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+compute_fpair(const F_FLOAT& rsq, const int& i, const int&j, const int& itype, const int& jtype) const {
+  (void) i;
+  (void) j;
+  const F_FLOAT r2inv = 1.0/rsq;
+  const int ljt = (STACKPARAMS?m_params[itype][jtype].lj_type:params(itype,jtype).lj_type);
+
+  const F_FLOAT lj_1 =  (STACKPARAMS?m_params[itype][jtype].lj1:params(itype,jtype).lj1);
+  const F_FLOAT lj_2 =  (STACKPARAMS?m_params[itype][jtype].lj2:params(itype,jtype).lj2);
+
+  /*if (ljt == LJ12_4) {
```
Comment (stanmoore1) [inline src/KOKKOS/pair_lj_spica_coul_long_kokkos.cpp]: @alphataubio can you please remove the commented out code, or justify why it is helpful?

File: src/KOKKOS/pair_lj_spica_coul_long_kokkos.cpp [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+template<bool STACKPARAMS,  class Specialisation>
+KOKKOS_INLINE_FUNCTION
+F_FLOAT PairLJSPICACoulLongKokkos<DeviceType>::
+compute_fcoul(const F_FLOAT& rsq, const int& /*i*/, const int&j,
+              const int& /*itype*/, const int& /*jtype*/,
+              const F_FLOAT& factor_coul, const F_FLOAT& qtmp) const {
+
+/*
+
+        r2inv = 1.0 / rsq;
```
Comment (stanmoore1) [inline src/KOKKOS/pair_lj_spica_coul_long_kokkos.cpp]: Please remove the commented out code, or justify why it is helpful

**Review by stanmoore1 — CHANGES_REQUESTED** [review]
A couple minor change requests, otherwise it looks great. Thanks @alphataubio!

**Comment by stanmoore1** [comment]
@alphataubio seeing a crash on V100: 

```
Exception: (CudaInternal::singleton().cuda_memcpy_async_wrapper( dst, src, n, cudaMemcpyDefault, s)) error( cudaErrorInvalidValue): invalid argument ../../lib/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:84
```

It runs correctly with UVM enabled, so probably a missing sync/modify.

File: src/KOKKOS/pair_lj_spica_coul_long_kokkos.cpp (hunk 1 above)
Comment (alphataubio) [reply src/KOKKOS/pair_lj_spica_coul_long_kokkos.cpp]: done

File: src/KOKKOS/pair_lj_spica_coul_long_kokkos.cpp (hunk 2 above)
Comment (alphataubio) [reply src/KOKKOS/pair_lj_spica_coul_long_kokkos.cpp]: done

File: src/KOKKOS/angle_spica_kokkos.cpp [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+    memoryKK->create_kokkos(k_eatom,eatom,maxeatom,"angle:eatom");
+    d_eatom = k_eatom.template view<DeviceType>();
+  }
+  if (vflag_atom) {
+    memoryKK->destroy_kokkos(k_vatom,vatom);
+    memoryKK->create_kokkos(k_vatom,vatom,maxvatom,"angle:vatom");
+    d_vatom = k_vatom.template view<DeviceType>();
+  }
+
+  k_k.template sync<DeviceType>();
```
Comment (stanmoore1) [inline src/KOKKOS/angle_spica_kokkos.cpp:90]: Need to do the same for other parameters like `k_repscale`, etc.

File: src/KOKKOS/angle_spica_kokkos.cpp [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+  f13 = e13 = delx3 = dely3 = delz3 = 0.0;
+
+  if (repflag) {
+
+    delx3 = x(i1,0) - x(i3,0);
+    dely3 = x(i1,1) - x(i3,1);
+    delz3 = x(i1,2) - x(i3,2);
+    const F_FLOAT rsq3 = delx3*delx3 + dely3*dely3 + delz3*delz3;
+
+    const int type1 = atom->type[i1];
```
Comment (stanmoore1) [inline src/KOKKOS/angle_spica_kokkos.cpp]: This is a pointer to host CPU memory and will segfault on GPUs. Need to use `atomKK->k_type` instead, after assigning the device view to a class copy, then use that in the device code.

File: src/KOKKOS/angle_spica_kokkos.cpp [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+    const int type3 = atom->type[i3];
+
+    f13=0.0;
+    e13=0.0;
+
+    if (rsq3 < rminsq[type1][type3]) {
```
Comment (stanmoore1) [inline src/KOKKOS/angle_spica_kokkos.cpp]: `rminsq`, `lj_type`, `lj1,2,3,4`, `emin` and any other array need to use Kokkos views.

**Review by stanmoore1 — COMMENTED** [review]
@alphataubio I think I found the issue, there are some references to CPU memory that will fail on GPUs, see the comments. The good news is that it gives the correct numerics with UVM enabled on GPUs, so you are close.

File: src/KOKKOS/angle_spica_kokkos.cpp (hunk 3 above)
Comment (alphataubio) [reply src/KOKKOS/angle_spica_kokkos.cpp:90]: i can work to finish fixing these next week. no rush to merge, that was my little pet project to participate in the workshop.

File: src/KOKKOS/angle_spica_kokkos.cpp (hunk 5 above)
Comment (stanmoore1) [reply src/KOKKOS/angle_spica_kokkos.cpp]: Just to clarify, you need to override `init_style()` from the parent `AngleSPICA`, call the parent version, then copy all the values into a Kokkos view.

**Comment by stanmoore1** [comment]
@alphataubio is this ready for review?

**Comment by alphataubio** [comment]
> @alphataubio is this ready for review?

no not even close. i added a bunch of views, sync's, modify's but i still get same error. it's not angle_spica_kokkos because when i run in.sds-regular with pair_lj_spica_kokkos instead of pair_lj_spica_coul_long_kokkos then it works fine without uvm. conclusion bug is in pair_lj_spica_coul_long_kokkos. cuda-gdb output:

```
Thread 1 "lmp_spica-kk" hit Breakpoint 2, Kokkos::Impl::cuda_internal_error_throw (e=cudaErrorInvalidValue, 
    name=0x11e49220 "(CudaInternal::singleton().cuda_memcpy_async_wrapper( dst, src, n, cudaMemcpyDefault, s))", file=0x11e490d0 "/dev/shm/lammps-alphataubio/lib/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp", 
    line=82) at /dev/shm/lammps-alphataubio/lib/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:172
172	  std::ostringstream out;
(cuda-gdb) bt
#0  Kokkos::Impl::cuda_internal_error_throw (e=cudaErrorInvalidValue, name=0x11e49220 "(CudaInternal::singleton().cuda_memcpy_async_wrapper( dst, src, n, cudaMemcpyDefault, s))", 
    file=0x11e490d0 "/dev/shm/lammps-alphataubio/lib/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp", line=82) at /dev/shm/lammps-alphataubio/lib/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:172
#1  0x0000000011cf4b0c in Kokkos::Impl::cuda_internal_safe_call (line=82, file=0x11e490d0 "/dev/shm/lammps-alphataubio/lib/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp", 
    name=0x11e49220 "(CudaInternal::singleton().cuda_memcpy_async_wrapper( dst, src, n, cudaMemcpyDefault, s))", e=<optimized out>)
    at /dev/shm/lammps-alphataubio/lib/kokkos/core/src/Cuda/Kokkos_Cuda_Error.hpp:64
#2  Kokkos::Impl::DeepCopyAsyncCuda (dst=0x7fff98200080, src=0x6201b8e80, n=1135824) at /dev/shm/lammps-alphataubio/lib/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:82
#3  0x0000000010d5e298 in Kokkos::Impl::DeepCopy<Kokkos::CudaHostPinnedSpace, Kokkos::CudaSpace, Kokkos::Serial, void>::DeepCopy (n=<optimized out>, src=0x6201b8e80, dst=0x7fff98200080, exec=..., 
    this=<optimized out>) at /dev/shm/lammps-alphataubio/lib/kokkos/core/src/Cuda/Kokkos_CudaSpace.hpp:496
#4  Kokkos::deep_copy<Kokkos::Serial, double* [3], Kokkos::LayoutRight, Kokkos::CudaHostPinnedSpace, Kokkos::MemoryTraits<1u>, double* [3], Kokkos::LayoutRight, Kokkos::Cuda, void>(Kokkos::Serial const&, Kokkos::View<double* [3], Kokkos::LayoutRight, Kokkos::CudaHostPinnedSpace, Kokkos::MemoryTraits<1u> > const&, Kokkos::View<double* [3], Kokkos::LayoutRight, Kokkos::Cuda, void> const&, std::enable_if<((Kokkos::is_execution_space<Kokkos::Serial>::value&&std::is_void<Kokkos::ViewTraits<double* [3], Kokkos::LayoutRight, Kokkos::CudaHostPinnedSpace, Kokkos::MemoryTraits<1u> >::specialize>::value)&&std::is_void<Kokkos::ViewTraits<double* [3], Kokkos::LayoutRight, Kokkos::Cuda, void>::specialize>::value)&&((((unsigned int)Kokkos::ViewTraits<double* [3], Kokkos::LayoutRight, Kokkos::CudaHostPinnedSpace, Kokkos::MemoryTraits<1u> >::rank)!=(0))||(((unsigned int)Kokkos::ViewTraits<double* [3], Kokkos::LayoutRight, Kokkos::Cuda, void>::rank)!=(0))), void>::type*) (exec_space=..., dst=..., src=...)
    at /dev/shm/lammps-alphataubio/lib/kokkos/core/src/Kokkos_CopyViews.hpp:2925
#5  0x0000000010d5f95c in LAMMPS_NS::AtomVecKokkos::perform_async_copy<Kokkos::DualView<double* [3], Kokkos::LayoutRight, Kokkos::Cuda, void> > (this=0x2ded3d50, src=..., space=<optimized out>)
    at /dev/shm/lammps-alphataubio/src/KOKKOS/atom_vec_kokkos.h:188
#6  0x0000000010e29e04 in LAMMPS_NS::AtomVecFullKokkos::sync_overlapping_device (this=<optimized out>, space=<optimized out>, mask=<optimized out>)
    at /dev/shm/lammps-alphataubio/src/KOKKOS/atom_vec_full_kokkos.cpp:1064
#7  0x0000000010316b44 in LAMMPS_NS::AtomKokkos::sync_overlapping_device (this=0x2deb7040, space=<optimized out>, mask=<optimized out>) at /dev/shm/lammps-alphataubio/src/KOKKOS/atom_kokkos.cpp:183
#8  0x0000000010761f64 in LAMMPS_NS::VerletKokkos::run (this=0x1fdc67e0, n=<optimized out>) at /dev/shm/lammps-alphataubio/src/KOKKOS/verlet_kokkos.cpp:441
#9  0x0000000010207378 in LAMMPS_NS::Run::command (this=0x2ded8bf0, narg=<optimized out>, arg=0x2dee6ad0) at /dev/shm/lammps-alphataubio/src/run.cpp:176
#10 0x00000000100a7818 in LAMMPS_NS::Input::execute_command (this=0x2d56d410) at /dev/shm/lammps-alphataubio/src/input.cpp:868
#11 0x00000000100a7c9c in LAMMPS_NS::Input::file (this=0x2d56d410) at /dev/shm/lammps-alphataubio/src/input.cpp:313
#12 0x000000001000f97c in main (argc=<optimized out>, argv=<optimized out>) at /dev/shm/lammps-alphataubio/src/main.cpp:77
(cuda-gdb) 

```

im now rewatching your recent KOKKOS youtube lecture to figure it out. however while i tested pair_lj_charmmfsw_coul_long_kokkos using in.charmmfsw, serial works fine but cuda gives random results within relative error ~1e-3 so based on your lecture that means i have a race condition somewhere in that class to solve also.

**Comment by stanmoore1** [comment]
@alphataubio it should be working now, it passes my test and gives the correct numerics on V100.
@akohlmey 8bba2d1 is a bugfix which needs to be backported to the stable.

File: src/KOKKOS/angle_spica_kokkos.cpp (hunk 3 above)
Comment (alphataubio) [reply src/KOKKOS/angle_spica_kokkos.cpp:90]: done

File: src/KOKKOS/angle_spica_kokkos.cpp (hunk 5 above)
Comment (alphataubio) [reply src/KOKKOS/angle_spica_kokkos.cpp]: done

**Comment by alphataubio** [comment]
> @alphataubio is this ready for review?

yes

**Review by stanmoore1 — APPROVED** [review]
I approve

**Pull request merged**